Introduction to GRPC - WIP (#237)
ippontech/blog-usa: posts/introduction-to-grpc.md

---
authors:
- Kenny Hegeland
  tags:
- gRPC
date: 2015-04-23T12:21:50.000Z
title: "Introduction to gRPC
image:
---

# What is gRPC?
gRPC (remote procedure call) is used for communicating between distributed systems (micro services).

> Google created and originally used an RPC framework called stubby, which later was open sourced as gRPC. 

Reviewer: ```suggestion
Google created and originally used an RPC framework called Stubby, which later was open sourced as gRPC. 
```

> While using protobuffers is not mandatory to use gRPC, it is generally the easiest approach (as it 

Reviewer: Could we have a brief description of a protobuf?

works out of the box). You define a protobuf in a simple message format and use protoc (the protobuf
compiler) to generate code to handle serializing and creating your domain objects. You can also create
a service in the protobuf definition which allows generation of methods for the client to communicate
with the server.

The protobuf simply defines the contract between the services, and this allows us to be language agnostic. 
Using libraries for specific language, you can generate language specific code and use a language specific
client to call the server. The server doesn’t need to know what language (this is also true of REST) the 
client is using, all it needs is the serialized message (which it knows how to deserialize).

Now that we’ve discussed what gRPC is, let’s flesh out an example. We will create a simple example 
application (available on GitHub - https://github.com/hegek87/derivative-solver) to compute the derivative 

> of simple polynomials. If you don’t remember how a derivative is calculated, we will quickly show the 

Reviewer: why are you using an example using polynomials? It couldn’t be something else?

`4*2 = 8`, then we subtract one from the exponent, `2-1=1`, giving us the derivative, `8x^1 = 8x`. 
This procedure works for simple polynomials, and can be expanded for polynomials of more than one term.
As another example, consider `x^3 + 3x^2 + 3`. Remember that x^0 = 1, so `3 = 3x^0`, and we can 
calculate the derivative as `3*0 = 0`. The full derivative of the equations leads to `3x^2 + 6x + 0 = 3x^2 + 6x`.

## Creating A Message Format
Now that we understand what a derivative is, let’s discuss how we represent a polynomial for our protobuf.
It seems like we have two simple ways to represent the polynomial, a naive way (which we handle in our rest
call) to do this is by sending it as a raw string. While this works, it could be cumbersome, as we will
need to find a way to split the string, parse the constants and exponents into numeric values and map back 
to a string after calculating the derivative. Perhaps a more elegant solution is to send the constant as an
array, and the exponents as an array, this way we can get the constant values and exponent values without 
any parsing on our side. Something like

```java

> Message Polynomial {

Reviewer: not sure if that was important, but seems that on your code you had the variables types too.

https://github.com/hegek87/derivative-solver/blob/8a2b7e3516ca2e9949592c36cde13127862215c0/server/src/main/proto/derivative.proto

```
message Polynomial {
  repeated int32 coefficients = 1;
  repeated int32 exponents = 2;
}
```

	repeated coefficients = 1;
	repeated exponents = 2;
}
```

Now that we have our message structure defined, we can compile to see the generated code. As expected, we
will see POJOs generated with setters, getters, and builders. Now that we've configured our message structure,
we can define our service method that will be called in order to actually make requests:
```java
service DerivativeSolver {
  rpc findDerivative(Polynomial) returns (Polynomial) {}
}
```

> Compiling using protoc once again, we will now have everything we need to set up our simple application. 

Reviewer: Further down you present some code, I would start here adding the command used to compile, so even a reader that don't know much about gRPC (like me) can compile your code.

        responseObserver.onError(err);
        throw err;
    }

    List<Integer> coefficients = polynomial.getCoefficientsList();
    List<Integer> exponents = polynomial.getExponentsList();

    List<Integer> updatedCoefficient = new ArrayList<>();
    List<Integer> updatedExponents = exponents.stream().map(exponent -> exponent - 1).collect(Collectors.toList());

    for(int i = 0; i < coefficients.size(); ++i) {
        updatedCoefficient.add(i, coefficients.get(i) * exponents.get(i));
    }

    responseObserver.onNext(Polynomial.newBuilder().addAllCoefficients(updatedCoefficient).addAllExponents(updatedExponents).build());
    responseObserver.onCompleted();
}
```

Setting this all up requires deploying both the server application, and the client application, both of which will use 
the same protobuf definitions. To run this locally, we've created simple docker files which do simple mapping of ports,
file systems and sets up a bridge network where they will communicate. In an actual deployed environment, we will 
likely have to use some sort of DNS to reference the host.

> Lots of code samples found in this article are snippets of our previously linked project which can be used as a 

Reviewer: ```suggestion
All the code samples from this article are snippets of our previously linked project which can be used as a 
```

> reference. The actual application made available has some other stuff built on top, following familiar spring patterns.

Reviewer: ```suggestion
reference. You can find on the application available on repository, shared on this article, some additional configurations related to the familiar spring patterns.
```

Among other things, we have a `@Configuration` class which is used to configure our gRPC server to be running inside 
our spring application. In this configuratioon, `DerivativeCalculatorService` houses our previously shown derivative 
logic. 

```java
@Configuration
public class GrpcServerConfiguration {

    private static final Logger LOGGER = LoggerFactory.getLogger(GrpcServerConfiguration.class);

    private int port;
    private Server server;
    private DerivativeCalculatorService derivativeCalculatorService;

    public GrpcServerConfiguration(ApplicationProperties applicationProperties, DerivativeCalculatorService derivativeCalculatorService) {
        this.port = applicationProperties.getGrpc().getPort();
        this.derivativeCalculatorService = derivativeCalculatorService;
    }

    @Bean
    public Server grpcServer() throws IOException {
        LOGGER.info("Starting GRPC Server");
        this.server = ServerBuilder.forPort(this.port).addService((BindableService) derivativeCalculatorService).build();
        this.server.start();
        return this.server;
    }
}
```

> Now we can finally try our example out. First we will want to build our two applications with maven, using the common 

Reviewer: You mean we can try it out using spring?
